runelite-client/src/main/java/net/runelite/client/plugins/grandexchange/GrandExchangePanel.java

```diff
@@ -0,0 +1,625 @@
+            );
+
+            //TODO: "empty slot" image [instead of dwarf remains image]?
+            imageLabels[idx].setIcon(offer.getItemId() == 0 ? new ImageIcon(ItemManager.getImage(0)) : offerImages[idx]);
+
+            itemNameLabels[idx].setText(offer.getItemId() == 0 ? "Empty" : offer.getTotalQuantity() + "x " + client.getItemDefinition(offer.getItemId()).getName());
+
+            offerPriceLabels[idx].setText(ItemManager.quantityToStackSize(offer.getPrice()).endsWith("M")
+                || ItemManager.quantityToStackSize(offer.getPrice()).endsWith("K")
+                ? ItemManager.quantityToStackSize(offer.getPrice())
+                : ItemManager.quantityToStackSize(offer.getPrice()) + " gp");
+
+            statusIndicators[idx].setText(getIsBuySell(offer));
+
+            EventQueue.invokeLater(() -> progressBars[idx].setValue(offer.getTotalQuantity() == 0 ? 0 : (offer.getQuantitySold() / offer.getTotalQuantity()) * 100));
+
+            progressBars[idx].setStringPainted(true);
+            progressBars[idx].setBackground(decodeColour(progressBars[idx]));
+            progressBars[idx].setForeground(Color.white);
+            progressBars[idx].setString(offer.getQuantitySold() + " / " + offer.getTotalQuantity());
+
+          });
+
+  private String getIsBuySell(GrandExchangeOffer offer)
+  {
+    switch (offer.getProgress()) {
+      case 2:
+        return "Buy";
+
+      case 5:
+        return "Bought";
+
+      case 10:
+        return "Sell";
+
+      case 13:
+        return "Sold";
+
+      case 0:
+      default:
+        return "";
+    }
+  }
+
+  private Color decodeColour(JProgressBar offerBar)
+  {
+    switch ((int) offerBar.getPercentComplete() * 100) {
+      case 0:
+        return Color.GRAY;
+
+      case 100:
+        return Color.GREEN;
+
+      default:
+        return Color.ORANGE;
+    }
+  }
+
+  private Function<GrandExchangeOffer[], String> getOfferStrings = (offers) ->
+  {
+    for (GrandExchangeOffer offer : offers) {
+      if (offer != null) {
+        return "GrandExchangeOffer{" +
+            "id=" + offer.getItemId() + ", " +
+            "price=" + offer.getPrice() + ", " +
+            "quantitySold=" + offer.getQuantitySold() + ", " +
+            "totalQuantity=" + offer.getTotalQuantity() +
+            "}";
+      }
+    }
+    return null;
+  };
+
+  private JComponent buildPanel()
+  {
+
+    JFrame frame = new JFrame();
+    frame.setBounds(100, 100, 225, 555);
+    frame.setDefaultCloseOperation(WindowConstants.EXIT_ON_CLOSE);
+
+    JPanel panel = new JPanel();
+    frame.getContentPane().add(panel, BorderLayout.CENTER);
+
+    JPanel panel_1 = new JPanel();
+    panel_1.setBackground(new Color(90, 82, 69));
+    panel_1.setBorder(new LineBorder(new Color(0, 0, 0)));
+
+    JPanel panel_2 = new JPanel();
+    panel_2.setBorder(new LineBorder(new Color(0, 0, 0)));
+    panel_2.setBackground(new Color(90, 82, 69));
+
+    imageLabels[1] = new JLabel("");
```

Comment:
This to line 240 looks duplicated 8 times. Consider moving the shared logic to a method.